jax/_src/random.py

```diff
@@ -840,8 +840,36 @@ def _dirichlet(key, alpha, shape, dtype):
     _check_shape("dirichlet", shape, np.shape(alpha)[:-1])
 
   alpha = lax.convert_element_type(alpha, dtype)
-  gamma_samples = gamma(key, alpha, shape + np.shape(alpha)[-1:], dtype)
-  return gamma_samples / jnp.sum(gamma_samples, axis=-1, keepdims=True)
+
+  # Compute gamma in log space, otherwise small alpha can lead to poor behavior.
+  log_gamma_samples = gamma(key, alpha, shape + np.shape(alpha)[-1:], dtype, log_space=True)
+  axes = (log_gamma_samples.ndim - 1,)
+  return lax.exp(log_gamma_samples - _logsumexp(log_gamma_samples, axes=axes, keepdims=True))
```

Comment:
Should we use `jax.nn.softmax` for this?

Reply:
Probably not, because we want the log-softmax, and computing the softmax first and then taking the log would lead re-introduce the same kinds of precision issues we're seeking to avoid.

Reply:
Done.